src/libraries/Microsoft.PowerFx.Connectors/PowerPlatformConnectorClient.cs

```diff
@@ -0,0 +1,92 @@
+        
+        /// <summary>
+        /// Session Id for telemetry.
+        /// </summary>
+        public string SessionId { get; set; } = Guid.NewGuid().ToString(); // "f4d37a97-f1c7-4c8c-80a6-f300c651568d"
+
+        private readonly HttpMessageInvoker _client;
+
+        public string ConnectionId { get; } 
+
+        // For example, "firstrelease-001.azure-apim.net" 
+        public string Endpoint { get; }
+
+        /// <summary>
+        /// Callback to get the auth token. 
+        /// Invoke as a callback since token may need to be refreshed. 
+        /// </summary>
+        public Func<string> GetAuthToken { get; }
+
+        public string EnvironmentId { get; set; }
+
+        public PowerPlatformConnectorClient(string endpoint, string environmentId, string connectionId, Func<string> getAuthToken, HttpMessageInvoker httpInvoker = null)
+        {
+            _client = httpInvoker ?? new HttpClient();
+
+            GetAuthToken = getAuthToken ?? throw new ArgumentNullException(nameof(getAuthToken));
+            ConnectionId = connectionId ?? throw new ArgumentNullException(nameof(connectionId));
+            EnvironmentId = environmentId ?? throw new ArgumentNullException(nameof(environmentId));
+            Endpoint = endpoint ?? throw new ArgumentNullException(nameof(endpoint));
+
+            // Must set to allow callers to invoke SendAsync() via other helper methods.
+            BaseAddress = new Uri("https://" + endpoint); // Uri.Parse will validate endpoint syntax. 
+        }
+                
+        public override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            var url = request.RequestUri.ToString();
+            if (request.RequestUri.IsAbsoluteUri)
+            {
+                // Client has Basepath set. 
+                // x-ms-request-url needs relative URL. 
+                throw new InvalidOperationException($"URL should be relative for x-ms-request-url property");
+            }
+
+            url = url.Replace("{connectionId}", ConnectionId);
+
+            var method = request.Method;
+            var useCache = method == HttpMethod.Get;
+
+            var authToken = GetAuthToken();
+
+            var req = new HttpRequestMessage(HttpMethod.Post, $"https://{Endpoint}/invoke");
+            req.Headers.Add("authority", Endpoint);
+            req.Headers.Add("scheme", "https");
+            req.Headers.Add("path", "/invoke");
+            req.Headers.Add("x-ms-client-session-id", SessionId);
+            req.Headers.Add("x-ms-request-method", method.ToString());
+            req.Headers.Add("authorization", "Bearer " + authToken);
+            req.Headers.Add("x-ms-client-environment-id", "/providers/Microsoft.PowerApps/environments/" + EnvironmentId);
+            req.Headers.Add("x-ms-user-agent", $"PowerFx/{Version}");
+
+            req.Headers.Add("x-ms-request-url", url); 
+
+            // $$$ Body? 
+
```

Comment:
What is the to-do here?

Reply:
handle body parameter which may need to be copied over from incoming request to APIM request.